lib/scala/project-manager/src/main/scala/org/enso/projectmanager/control/core/Applicative.scala

```diff
@@ -0,0 +1,43 @@
+package org.enso.projectmanager.control.core
+
+import zio.ZIO
+
+/**
+  * An applicative functor.
+  */
+trait Applicative[F[+_, +_]] {
+
+  /**
+    * Lifts value into the applicative functor.
+    */
+  def pure[E, A](a: A): F[E, A]
+
+  /**
+    * Applies the function in the applicative context to the given value.
+    */
+  def ap[E, E1 >: E, A, B](ff: F[E, A => B])(fa: F[E1, A]): F[E1, B]
+
+  /**
+    * Applies function f to the given value.
+    */
+  def map[E, A, B](fa: F[E, A])(f: A => B): F[E, B]
+
+  /**
+    * Applies the binary function to the effectful falues ma and mb.
+    */
+  def map2[E, A, B, C](ma: F[E, A], mb: F[E, B])(f: (A, B) => C): F[E, C]
+}
+
+object Applicative {
+
+  def apply[F[+_, +_]](implicit applicative: Applicative[F]): Applicative[F] =
+    applicative
+
+  implicit def zioApplicative[R]: Applicative[ZIO[R, +*, +*]] =
+    zioApplicativeInstance.asInstanceOf[Applicative[ZIO[R, +*, +*]]]
+
+  final private[this] val zioApplicativeInstance
+    : Applicative[ZIO[Any, +*, +*]] =
+    new ZioApplicative[Any]
```

Comment:
I assume that this cast is safe because of type erasure, but wouldn't it be better to just create a `new` instance with the right type inside of the `implicit def`?

Reply:
AFAIK is a trick to initialize the class once, instead of creating a new one each time the implicit function is called.

Reply:
Makes sense, although I wonder how big of a performance impact that has.
I think it may be worth at least adding a short comment about its correctness, because it goes around the type-system a bit.

Reply:
I think it's ok to leave it like that because it repeats itself in the codebase and forms a pattern. Also, it's not hard to see how `zioApplicative` function and `zioApplicativeConstant` are interacting. Too many comments is also not a good thing.